### PR TITLE
adds lookupCastConversation, updates searchUser params, and depreciates fetchAllCastsInThread

### PR DESCRIPTION
**Pull request by MrKevinOConnell**
New Method

Cast

lookupCastConversation -> https://api.neynar.com/v2/farcaster/cast/conversation

Updates Existing Methods

User
Add limit and cursor in searchUser, make viewerFid optional -> https://docs.neynar.com/reference/user-search

Depreciated

fetchAllCastsInThread -> https://docs.neynar.com/reference/all-casts-in-thread-v1

### CODE REVIEW

File: src/neynar-api/v2/client.ts [hunk 1]
```diff
@@ -927,6 +928,48 @@ export class NeynarV2APIClient {
...
+   * ).then(response => {
+   *   console.log('Bulk Casts Information:', response); // Outputs information about the specified casts
+   * });
+   *
+   * For more information, refer to the [Neynar documentation](https://docs.neynar.com/reference/cast-conversation).
+   */
+    public async lookupCastConversation(
+      castHashOrUrl: string,
+      type: CastParamType,
+      replyDepth?: number,
```
Comment (Shreyaschorge) [inline src/neynar-api/v2/client.ts:962]: replyDepth must be a key-value argument. We follow this practice where all the optional args are KV args

File: src/neynar-api/v2/client.ts [hunk 2]
```diff
@@ -927,6 +928,48 @@ export class NeynarV2APIClient {
...
+   * client.lookupCastConversation(https://warpcast.com/rish/0x9288c1,
+   *  type: 'url',
+   *  replyDepth: 2
+   * ).then(response => {
+   *   console.log('Bulk Casts Information:', response); // Outputs information about the specified casts
+   * });
+   *
+   * For more information, refer to the [Neynar documentation](https://docs.neynar.com/reference/cast-conversation).
+   */
+    public async lookupCastConversation(
```
Comment (Shreyaschorge) [inline src/neynar-api/v2/client.ts:959]: `lookupCastConversation` should also be present in `src/neynar-api/neynar-api-client.ts` and this method should be called from there.

If not present then, it won't show up in sdk

File: src/neynar-api/v2/client.ts [hunk 3]
```diff
@@ -927,6 +928,48 @@ export class NeynarV2APIClient {
     return response.data;
   }
 
+    /**
+   * Retrieves information about multiple casts using an array of their hashes. This method is useful
+   * for fetching details of several casts at once, identified by their unique hashes. Optional parameters
+   * allow for adding viewer context to the cast objects to show whether the viewer has liked or recasted
+   * the cast and sorting the casts based on different criteria.
+   *
+   * @param {Array<string>} castsHashes - An array of strings representing the hashes of the casts
```
Comment (Shreyaschorge) [inline src/neynar-api/v2/client.ts:937]: The documentation for args doesn't match the method args.

castsHashes, viewerFid, and sortType is not accepted by `lookupCastConversation`

File: src/neynar-api/v2/client.ts [hunk 4]
```diff
@@ -927,6 +928,48 @@ export class NeynarV2APIClient {
...
+   *
+   * @returns {Promise<CastsResponse>} A promise that resolves to a `CastsResponse` object,
+   *   containing information about the requested casts.
+   *
+   * @example
+   * // Example: Fetch information about a casts conversation using a given hash and type
+   *
+   *
+   * client.lookupCastConversation(https://warpcast.com/rish/0x9288c1,
+   *  type: 'url',
```
Comment (Shreyaschorge) [inline src/neynar-api/v2/client.ts:951]: CastParamType.Url should be used here. We are doing this so new users can get an idea of how to use our SDK

Search for `CastParamType` in this file, you'll get an example on how to write the examples for the enums

File: src/neynar-api/v2/client.ts [hunk 5]
```diff
@@ -927,6 +928,48 @@ export class NeynarV2APIClient {
...
+   *  replyDepth: 2
+   * ).then(response => {
+   *   console.log('Bulk Casts Information:', response); // Outputs information about the specified casts
```
Comment (Shreyaschorge) [inline src/neynar-api/v2/client.ts:954]: Log can be updated to `Cast Conversation`

File: src/neynar-api/v2/client.ts [hunk 6]
```diff
@@ -927,6 +928,48 @@ export class NeynarV2APIClient {
...
+   * allow for adding viewer context to the cast objects to show whether the viewer has liked or recasted
+   * the cast and sorting the casts based on different criteria.
+   *
+   * @param {Array<string>} castsHashes - An array of strings representing the hashes of the casts
+   *   to be retrieved.
+   * @param {Object} [options] - Optional parameters for the request.
+   * @param {number} [options.viewerFid] - Adds viewer context to cast objects to indicate whether the viewer has liked or recasted the cast.
+   * @param {'trending' | 'likes' | 'recasts' | 'replies' | 'recent'} [options.sortType] - Optional parameter to sort the casts based on different criteria such as trending, likes, recasts, replies, or recent.
+   *
+   * @returns {Promise<CastsResponse>} A promise that resolves to a `CastsResponse` object,
```
Comment (Shreyaschorge) [inline src/neynar-api/v2/client.ts:943]: Method returns `Promise<ConversationContainer>`

File: src/neynar-api/v2/client.ts [hunk 7]
```diff
@@ -927,6 +928,48 @@ export class NeynarV2APIClient {
...
+   *   console.log('Bulk Casts Information:', response); // Outputs information about the specified casts
+   * });
+   *
+   * For more information, refer to the [Neynar documentation](https://docs.neynar.com/reference/cast-conversation).
+   */
+    public async lookupCastConversation(
+      castHashOrUrl: string,
+      type: CastParamType,
+      replyDepth?: number,
+    ): Promise<ConversationContainer> {
```
Comment (Shreyaschorge) [inline src/neynar-api/v2/client.ts:963]: Not sure if we should call this schema ConversationContainer

@manan19 Thoughts..??

File: src/neynar-api/v2/client.ts [hunk 8]
```diff
@@ -927,6 +928,48 @@ export class NeynarV2APIClient {
     return response.data;
   }
 
+    /**
+   * Retrieves information about multiple casts using an array of their hashes. This method is useful
```
Comment (Shreyaschorge) [inline src/neynar-api/v2/client.ts:932]: This description is for `fetchBulkCasts`. Please add suitable description for `lookupCastConversation`

File: src/neynar-api/v2/openapi-farcaster/apis/user-api.ts [hunk 9]
```diff
@@ -593,17 +593,17 @@ export const UserApiAxiosParamCreator = function (configuration?: Configuration)
...
          * @param {string} q 
-         * @param {number} viewerFid 
+         * @param {number} [viewerFid] 
+         * @param {number} [limit] 
+         * @param {string} [cursor] Pagination cursor.
          * @param {*} [options] Override http request option.
          * @throws {RequiredError}
          */
-        userSearch: async (apiKey: string, q: string, viewerFid: number, options: AxiosRequestConfig = {}): Promise<RequestArgs> => {
+        userSearch: async (apiKey: string, q: string, viewerFid?: number, limit?: number, cursor?: string, options: AxiosRequestConfig = {}): Promise<RequestArgs> => {
```
Comment (Shreyaschorge) [inline src/neynar-api/v2/openapi-farcaster/apis/user-api.ts:602]: Please reflect these changes in `searchUser` method in `src/neynar-api/v2/client.ts` and `src/neynar-api/neynar-api-client.ts`

also update the JS doc string for those `searchUser`

**Review by Shreyaschorge — CHANGES_REQUESTED** [review]

File: src/neynar-api/v2/client.ts [hunk 10]
```diff
@@ -927,6 +932,47 @@ export class NeynarV2APIClient {
...
+ * @param {string} castHashOrUrl - The hash or URL of the cast conversation to be retrieved.
+ * @param {CastParamType} type - Specifies the type of the provided identifier, indicating whether it's a hash or URL.
+ * @param {number} [replyDepth] - Optional parameter to specify how deep the reply chain should be fetched.
+ *
+ * @returns {Promise<ConversationContainer>} A promise that resolves to a `ConversationContainer` object,
+ *   containing detailed information about the requested cast conversation, including replies up to the specified depth.
+ *
+ * @example
+ * // Fetch detailed information about a cast conversation using a given hash or URL, with a reply depth of 2
+ * client.lookupCastConversation(
```
Comment (Shreyaschorge) [inline src/neynar-api/v2/client.ts:950]: Add an import statement as well so users can simply copy-paste it

import { CastParamType } from "@neynar/nodejs-sdk";

File: src/neynar-api/neynar-api-client.ts [hunk 11]
```diff
@@ -1251,10 +1252,49 @@ export class NeynarAPIClient {
...
+ * @param {CastParamType} type - Enumerates the identifier type, specifying whether the provided value is a hash or a URL, thus guiding the retrieval process appropriately.
+ * @param {Object} [options] - An optional parameter object to refine the query.
+ * @param {number} [options.replyDepth] - An optional parameter within the options object, specifying the desired depth of replies to fetch within the conversation. This allows for tailored retrieval of conversation data, ranging from top-level casts only to deeper, more comprehensive conversation threads.
+ *
+ * @returns {Promise<ConversationContainer>} A promise resolving to a `ConversationContainer` object. This object encapsulates detailed information about the cast conversation, including the content of the conversation itself and any replies, structured up to the specified depth.
+ *
+ * @example
+ * // Example usage: Retrieve detailed information about a cast conversation via URL, including replies up to two levels deep
+ * client.lookupCastConversation(
```
Comment (Shreyaschorge) [inline src/neynar-api/neynar-api-client.ts:1273]: Add an import statement as well so users can simply copy-paste it

import { CastParamType } from "@neynar/nodejs-sdk";

File: src/neynar-api/neynar-api-client.ts [hunk 12]
```diff
@@ -1251,10 +1252,49 @@ export class NeynarAPIClient {
    */
   public async searchUser(
     q: string,
-    viewerFid: number
+    viewerFid: number,
+    options?: { limit?: number, cursor?: string}
```
Comment (Shreyaschorge) [inline src/neynar-api/neynar-api-client.ts:1256]: Adding a limit in the example would be nice. So new users would know how to pass optional params.

**Review by Shreyaschorge — CHANGES_REQUESTED** [review]
Minor changes then it should be good to go

File: src/neynar-api/v2/client.ts (hunk 7 above)
Comment (manan19) [reply src/neynar-api/v2/client.ts]: If `Conversation` isn't used, we should just use it

Comment (MrKevinOConnell) [reply src/neynar-api/v2/client.ts]: @manan19 does this mean the oas needs to be updated also?

Comment (Shreyaschorge) [reply src/neynar-api/v2/client.ts]: @MrKevinOConnell Yes, after generating code and the wrapper code will change too.

File: src/neynar-api/neynar-api-client.ts [hunk 13]
```diff
@@ -1243,18 +1244,59 @@ export class NeynarAPIClient {
...
+   * client.searchUser('ris', 19960, { limit: 10 }).then(response => {
    *   console.log('User Search Results:', response); // Outputs the results of the user search
    * });
    *
    * For more information, refer to the [Neynar documentation](https://docs.neynar.com/reference/user-search).
    */
   public async searchUser(
     q: string,
-    viewerFid: number
+    viewerFid: number,
```
Comment (Shreyaschorge) [inline src/neynar-api/neynar-api-client.ts:1255]: @MrKevinOConnell `viewerFid` is optional now, but please do not make it a part of KV args (it'll be a breaking change).

For now, make it optional, we'll address these breaking changes in the next major release.

**Review by Shreyaschorge — CHANGES_REQUESTED** [review]
Minor change.

File: src/neynar-api/neynar-api-client.ts (hunk 13 above)
Comment (MrKevinOConnell) [reply src/neynar-api/neynar-api-client.ts:1255]: sounds good!

File: src/neynar-api/v2/client.ts [hunk 14]
```diff
@@ -927,6 +932,50 @@ export class NeynarV2APIClient {
...
+ * @param {boolean} [includeChronologicalParentCasts] - Optional parameter to include chronological parent casts in the response.
+ * @returns {Promise<Conversation>} A promise that resolves to a `Conversation` object,
+ *   containing detailed information about the requested cast conversation, including replies up to the specified depth.
+ *
+ * @example
+ * // Fetch detailed information about a cast conversation using a given hash or URL, with a reply depth of 2
+ * client.lookupCastConversation(
+ *   'https://warpcast.com/rish/0x9288c1',
+ *   CastParamType.Url,
+ *   2,
```
Comment (Shreyaschorge) [inline src/neynar-api/v2/client.ts:953]: @MrKevinOConnell Optional parameters are not passed correctly it should be `{ replyDepth: 2, includeChronologicalParentCasts: true}`

Though this example will not be displayed anywhere rn, maybe in the future if we decide to expose version-specific methods then this example will be incorrect.

**Review by Shreyaschorge — CHANGES_REQUESTED** [review]
The rest looks good

**Review by Shreyaschorge — APPROVED** [review]
LGTM

**Pull request merged**